[java] Use Process class to execute SM in Java (fix #13091)

bonigarcia opened the pull request:
### Description
This PR changes the use of the `org.openqa.selenium.build.ExternalProcess` Selenium class to the standard `java.lang.Process` (through `java.lang.ProcessBuilder`) to execute Selenium Manager in Java.

UPDATE: The second commit in this PR includes some logic for retrying the command execution, as reported in #13091.

### Motivation and Context
It has been reported by some users (see #13091) that sometimes, the `SeleniumManager` Java logic fails due to an empty output by Selenium Manager. After troubleshooting it, it seems the cause is the use of `ExternalProcess`, and using `Process` fixes it.

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] I have read the [contributing](https://github.com/SeleniumHQ/selenium/blob/trunk/CONTRIBUTING.md) document.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
<!--- Provide a general summary of your changes in the Title above -->

joerg1985 commented:
I think it might be better to fix the root cause in the ExternalProcess class.

One possible cause could be reading the outputstream asynchronous in a different thread.
If the thread is not started yet / has not finished to copy the output the getOutput might be incomplete.

A possible fix to this would be to call `Thread.join` for the copy thread when `ExternalProcess.waitFor` has been called.

I could create a PR for this.

bonigarcia commented:
I agree that if the `ExternalProcess` is used in another place of the Selenium code base, it'd be better to be fixed.

But in the particular case of the `SeleniumManager` class. What is the benefit of using `ExternalProcess` vs `Process`? I suppose I don't see the whole picture. But if both implementations work the same way, wouldn't it be better to rely on the standard Java library than on our proper implementation?

joerg1985 commented:
The `ExternalProcess` does ensure you don't hit common issues with the Process class and should not add new ones ;)

One example, if the SM would send a huge output to stdout the process would lock as soon as the outputbuffer is full.
The output must be read to ensure the process can further write to stdout.

In the PR the java code will wait for the process to exit, before reading the stdout.
As soon as the outputbuffer is full, you have a kind of deadlock situation.

bonigarcia commented:
Just for the record, the SM output should not be long. It should be a JSON file with around 55 lines (i.e., less than 2K). In case of an unexpected error, the output (panic description) should be even less.

joerg1985 commented:
I just pushed d1787a9d1ed6a3ea643e9d50f393fec02cc6ae2e to ensure processing the output has been completed before returning from a `.shutdown` or `.waitFor` call.

dummyuser992 commented:
are there any updates here?

diemol commented:
@bonigarcia @joerg1985 did you folks agree on something? Are we using the `ExternalProcess` class or not?

joerg1985 commented:
It is safe to use the `ExternalProcess` now. The PR does contain a retry on Exception mechanism as a workaround to the `error=26, Text file busy.`, this should be rebased in my mind.

joerg1985 commented:
PS: https://github.com/SeleniumHQ/docker-selenium/issues/2008 should make the retry obsolete and it might be a good idea to remove the retry again, to see if the root issue happens outside docker too.

diemol commented:
> PS: [SeleniumHQ/docker-selenium#2008](https://github.com/SeleniumHQ/docker-selenium/issues/2008) should make the retry obsolete and it might be a good idea to remove the retry again, to see if the root issue happens outside docker too.

I don't remember where, but someone mentioned this issue outside Docker.

bonigarcia commented:
I don't have a strong opinion on this. I created this PR for the first time because there was a problem related to the execution of Selenium Manager in Java, in which the Selenium Manager was sometimes empty. My intuition was the problem in which the SM process was executed in Java. For that reason, I changed using the `org.openqa.selenium.build.ExternalProcess` Selenium class to the standard `java.lang.Process` in the first commit of this PR. Thanks to this change, we ensured the problem was not on the Rust side but on the Java side.

After solving the first issue, a second problem occurs (`error=26, Text file busy`). Since there was no solution for that problem, I added a second commit based on a retries mechanism. That workaround seems to work nicely; at least, users have reported that it works in #13091. If you prefer to solve this problem another way, I'm okay with that. But If you discard this PR, it would be good to test the new approach properly.

Regarding the use of `Process` vs. `ExternalProcess`, in my experience, the standard Java class `Process` works nicely. I used it for years in WebDriverManager and never had problems with it. The problem with a long output by Selenium Manager should not happen, so I don't see any particular value in using `ExternalProcess` in the `SeleniumManager` Java class. But if you are 100% sure that `ExternalProcess` is correct now, go ahead and use it.

krmahadevan commented:
@bonigarcia - [ExternalProcess](https://github.com/SeleniumHQ/selenium/blob/trunk/java/src/org/openqa/selenium/os/ExternalProcess.java) I think is just an abstraction of `Process`. I see that it internally leverages the `Process` class to get the work done.

diemol commented:
I released a new set of Docker images (https://github.com/SeleniumHQ/docker-selenium/releases/tag/4.15.0-20231122), now the browser driver path is set via properties and with that, Selenium Manager is avoided. 

I would leave this PR open for a while in case people still report this issue in non Docker environments.

ctunkl commented:
Hi there!

> > PS: [SeleniumHQ/docker-selenium#2008](https://github.com/SeleniumHQ/docker-selenium/issues/2008) should make the retry obsolete and it might be a good idea to remove the retry again, to see if the root issue happens outside docker too.
> 
> I don't remember where, but someone mentioned this issue outside Docker.

That wasn't me, but we do have this problem outside of Docker. It's the same "error failed to parse json output" error mentioned in https://github.com/SeleniumHQ/selenium/issues/13091:

`net.thucydides.core.webdriver.DriverConfigurationError: WebDriver was unable to create a new instance of type class org.openqa.selenium.chrome.ChromeDriver
WebDriver reported the following message: Unable to obtain: Capabilities {acceptInsecureCerts: false, browserName: chrome, goog:chromeOptions: {args: [remote-allow-origins=*, --headless=new, --window-size=1200,1600], extensions: []}}, error Failed to parse json output, executed: [--browser, chrome, --output, json]`

diemol commented:
> Hi there!
> 
> > > PS: [SeleniumHQ/docker-selenium#2008](https://github.com/SeleniumHQ/docker-selenium/issues/2008) should make the retry obsolete and it might be a good idea to remove the retry again, to see if the root issue happens outside docker too.
> > 
> > 
> > I don't remember where, but someone mentioned this issue outside Docker.
> 
> That wasn't me, but we do have this problem outside of Docker. It's the same "error failed to parse json output" error mentioned in #13091:
> 
> `net.thucydides.core.webdriver.DriverConfigurationError: WebDriver was unable to create a new instance of type class org.openqa.selenium.chrome.ChromeDriver WebDriver reported the following message: Unable to obtain: Capabilities {acceptInsecureCerts: false, browserName: chrome, goog:chromeOptions: {args: [remote-allow-origins=*, --headless=new, --window-size=1200,1600], extensions: []}}, error Failed to parse json output, executed: [--browser, chrome, --output, json]`

Have you tried again with the Nightly version? https://www.selenium.dev/downloads/#nightly

ctunkl commented:
> > Hi there!
> > > > PS: [SeleniumHQ/docker-selenium#2008](https://github.com/SeleniumHQ/docker-selenium/issues/2008) should make the retry obsolete and it might be a good idea to remove the retry again, to see if the root issue happens outside docker too.
> > > 
> > > 
> > > I don't remember where, but someone mentioned this issue outside Docker.
> > 
> > 
> > That wasn't me, but we do have this problem outside of Docker. It's the same "error failed to parse json output" error mentioned in #13091:
> > `net.thucydides.core.webdriver.DriverConfigurationError: WebDriver was unable to create a new instance of type class org.openqa.selenium.chrome.ChromeDriver WebDriver reported the following message: Unable to obtain: Capabilities {acceptInsecureCerts: false, browserName: chrome, goog:chromeOptions: {args: [remote-allow-origins=*, --headless=new, --window-size=1200,1600], extensions: []}}, error Failed to parse json output, executed: [--browser, chrome, --output, json]`
> 
> Have you tried again with the Nightly version? https://www.selenium.dev/downloads/#nightly

I have done so now (using 4.16.0-SNAPSHOT), and the problem seems to be fixed! 🎉

diemol commented:
> > > Hi there!
> > > > > PS: [SeleniumHQ/docker-selenium#2008](https://github.com/SeleniumHQ/docker-selenium/issues/2008) should make the retry obsolete and it might be a good idea to remove the retry again, to see if the root issue happens outside docker too.
> > > > 
> > > > 
> > > > I don't remember where, but someone mentioned this issue outside Docker.
> > > 
> > > 
> > > That wasn't me, but we do have this problem outside of Docker. It's the same "error failed to parse json output" error mentioned in #13091:
> > > `net.thucydides.core.webdriver.DriverConfigurationError: WebDriver was unable to create a new instance of type class org.openqa.selenium.chrome.ChromeDriver WebDriver reported the following message: Unable to obtain: Capabilities {acceptInsecureCerts: false, browserName: chrome, goog:chromeOptions: {args: [remote-allow-origins=*, --headless=new, --window-size=1200,1600], extensions: []}}, error Failed to parse json output, executed: [--browser, chrome, --output, json]`
> > 
> > 
> > Have you tried again with the Nightly version? https://www.selenium.dev/downloads/#nightly
> 
> I have done so now (using 4.16.0-SNAPSHOT), and the problem seems to be fixed! 🎉

Great. Then it seems this PR won't be needed. Let's wait until we get more feedback.

giuliohome commented:
> But If you discard this PR, it would be good to test the new approach properly.

@bonigarcia  💯 agree with you. I'd especially appreciate it if someone could [include this test as well](https://github.com/SeleniumHQ/selenium/issues/12034#issuecomment-1790649468) among the others. Well, in the meantime, I've already confirmed that the download test is okay in Docker. I'll now try it in a real Kubernetes cluster. Tested also in AKS, [it seems ok](https://github.com/SeleniumHQ/selenium/issues/13091#issuecomment-1825476477).

giuliohome commented:
As a follow up to my previous comment, I paste here a log that I see in selenium router a few hours after the upgrade

```
2023-11-28T11:37:40+01:00	10:37:40.989 WARN [SeleniumSpanExporter$1.lambda$export$3] - {"traceId": "660c77388fb6b400b7095b2d7272903a","eventTime": 1701167860983161014,"eventName": "exception","attributes": {"exception.message": "Unable to execute request: java.util.concurrent.TimeoutException\nBuild info: version: '4.15.0', revision: '1d14b5521b'\nSystem info: os.name: 'Linux', os.arch: 'amd64', os.version: '5.15.0-1036-azure', java.version: '11.0.20.1'\nDriver info: driver.version: unknown","exception.stacktrace": "org.openqa.selenium.TimeoutException: java.util.concurrent.TimeoutException\nBuild info: version: '4.15.0', revision: '1d14b5521b'\nSystem info: os.name: 'Linux', os.arch: 'amd64', os.version: '5.15.0-1036-azure', java.version: '11.0.20.1'\nDriver info: driver.version: unknown\n\tat org.openqa.selenium.remote.http.jdk.JdkHttpClient.execute0(JdkHttpClient.java:410)\n\tat org.openqa.selenium.remote.http.AddSeleniumUserAgent.lambda$apply$0(AddSeleniumUserAgent.java:42)\n\tat org.openqa.selenium.remote.http.Filter.lambda$andFinally$1(Filter.java:55)\n\tat org.openqa.selenium.remote.http.jdk.JdkHttpClient.execute(JdkHttpClient.java:366)\n\tat org.openqa.selenium.remote.tracing.TracedHttpClient.execute(TracedHttpClient.java:54)\n\tat org.openqa.selenium.grid.security.AddSecretFilter.lambda$apply$0(AddSecretFilter.java:40)\n\tat org.openqa.selenium.remote.http.Filter.lambda$andFinally$1(Filter.java:55)\n\tat org.openqa.selenium.grid.sessionqueue.remote.RemoteNewSessionQueue.addToQueue(RemoteNewSessionQueue.java:104)\n\tat org.openqa.selenium.grid.sessionqueue.NewSessionQueue.lambda$new$0(NewSessionQueue.java:68)\n\tat org.openqa.selenium.remote.http.Route$TemplatizedRoute.handle(Route.java:193)\n\tat org.openqa.selenium.remote.http.Route.execute(Route.java:69)\n\tat org.openqa.selenium.remote.http.Route$CombinedRoute.handle(Route.java:345)\n\tat org.openqa.selenium.remote.http.Route.execute(Route.java:69)\n\tat org.openqa.selenium.grid.sessionqueue.NewSessionQueue.execute(NewSessionQueue.java:128)\n\tat org.openqa.selenium.remote.tracing.SpanWrappedHttpHandler.execute(SpanWrappedHttpHandler.java:87)\n\tat org.openqa.selenium.remote.http.Filter$1.execute(Filter.java:63)\n\tat org.openqa.selenium.remote.http.Route$CombinedRoute.handle(Route.java:345)\n\tat org.openqa.selenium.remote.http.Route.execute(Route.java:69)\n\tat org.openqa.selenium.grid.router.Router.execute(Router.java:87)\n\tat org.openqa.selenium.grid.web.EnsureSpecCompliantResponseHeaders.lambda$apply$0(EnsureSpecCompliantResponseHeaders.java:34)\n\tat org.openqa.selenium.remote.http.Filter$1.execute(Filter.java:63)\n\tat org.openqa.selenium.remote.http.Route$CombinedRoute.handle(Route.java:345)\n\tat org.openqa.selenium.remote.http.Route.execute(Route.java:69)\n\tat org.openqa.selenium.remote.http.Route$NestedRoute.handle(Route.java:271)\n\tat org.openqa.selenium.remote.http.Route.execute(Route.java:69)\n\tat org.openqa.selenium.remote.http.Route$CombinedRoute.handle(Route.java:345)\n\tat org.openqa.selenium.remote.http.Route.execute(Route.java:69)\n\tat org.openqa.selenium.remote.http.Route$CombinedRoute.handle(Route.java:345)\n\tat org.openqa.selenium.remote.http.Route.execute(Route.java:69)\n\tat org.openqa.selenium.remote.http.Route$CombinedRoute.handle(Route.java:345)\n\tat org.openqa.selenium.remote.http.Route.execute(Route.java:69)\n\tat org.openqa.selenium.remote.http.Route$CombinedRoute.handle(Route.java:345)\n\tat org.openqa.selenium.remote.http.Route.execute(Route.java:69)\n\tat org.openqa.selenium.remote.AddWebDriverSpecHeaders.lambda$apply$0(AddWebDriverSpecHeaders.java:35)\n\tat org.openqa.selenium.remote.ErrorFilter.lambda$apply$0(ErrorFilter.java:44)\n\tat org.openqa.selenium.remote.http.Filter$1.execute(Filter.java:63)\n\tat org.openqa.selenium.remote.ErrorFilter.lambda$apply$0(ErrorFilter.java:44)\n\tat org.openqa.selenium.remote.http.Filter$1.execute(Filter.java:63)\n\tat org.openqa.selenium.netty.server.SeleniumHandler.lambda$channelRead0$0(SeleniumHandler.java:44)\n\tat java.base\u002fjava.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)\n\tat java.base\u002fjava.util.concurrent.FutureTask.run(FutureTask.java:264)\n\tat java.base\u002fjava.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)\n\tat java.base\u002fjava.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)\n\tat java.base\u002fjava.lang.Thread.run(Thread.java:829)\nCaused by: java.util.concurrent.TimeoutException\n\tat java.base\u002fjava.util.concurrent.CompletableFuture.timedGet(CompletableFuture.java:1886)\n\tat java.base\u002fjava.util.concurrent.CompletableFuture.get(CompletableFuture.java:2021)\n\tat org.openqa.selenium.remote.http.jdk.JdkHttpClient.execute0(JdkHttpClient.java:393)\n\t... 43 more\n","exception.type": "org.openqa.selenium.TimeoutException","http.client_ip": "10.65.131.136","http.flavor": 1,"http.handler_class": "org.openqa.selenium.grid.sessionqueue.remote.RemoteNewSessionQueue","http.host": "test-auto.eni.com","http.method": "POST","http.request_content_length": "75","http.scheme": "HTTP","http.target": "\u002fsession","http.user_agent": "selenium\u002f4.15.0 (js linux)"}}
2023-11-28T11:37:40+01:00	
2023-11-28T11:37:40+01:00		... 43 more
2023-11-28T11:37:40+01:00		at org.openqa.selenium.remote.http.jdk.JdkHttpClient.execute0(JdkHttpClient.java:393)
2023-11-28T11:37:40+01:00		at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2021)
2023-11-28T11:37:40+01:00		at java.base/java.util.concurrent.CompletableFuture.timedGet(CompletableFuture.java:1886)
2023-11-28T11:37:40+01:00	Caused by: java.util.concurrent.TimeoutException
2023-11-28T11:37:40+01:00		at java.base/java.lang.Thread.run(Thread.java:829)
2023-11-28T11:37:40+01:00		at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
2023-11-28T11:37:40+01:00		at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
2023-11-28T11:37:40+01:00		at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
2023-11-28T11:37:40+01:00		at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.netty.server.SeleniumHandler.lambda$channelRead0$0(SeleniumHandler.java:44)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.remote.http.Filter$1.execute(Filter.java:63)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.remote.ErrorFilter.lambda$apply$0(ErrorFilter.java:44)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.remote.http.Filter$1.execute(Filter.java:63)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.remote.ErrorFilter.lambda$apply$0(ErrorFilter.java:44)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.remote.AddWebDriverSpecHeaders.lambda$apply$0(AddWebDriverSpecHeaders.java:35)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.remote.http.Route.execute(Route.java:69)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.remote.http.Route$CombinedRoute.handle(Route.java:345)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.remote.http.Route.execute(Route.java:69)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.remote.http.Route$CombinedRoute.handle(Route.java:345)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.remote.http.Route.execute(Route.java:69)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.remote.http.Route$CombinedRoute.handle(Route.java:345)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.remote.http.Route.execute(Route.java:69)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.remote.http.Route$CombinedRoute.handle(Route.java:345)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.remote.http.Route.execute(Route.java:69)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.remote.http.Route$NestedRoute.handle(Route.java:271)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.remote.http.Route.execute(Route.java:69)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.remote.http.Route$CombinedRoute.handle(Route.java:345)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.remote.http.Filter$1.execute(Filter.java:63)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.grid.web.EnsureSpecCompliantResponseHeaders.lambda$apply$0(EnsureSpecCompliantResponseHeaders.java:34)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.grid.router.Router.execute(Router.java:87)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.remote.http.Route.execute(Route.java:69)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.remote.http.Route$CombinedRoute.handle(Route.java:345)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.remote.http.Filter$1.execute(Filter.java:63)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.remote.tracing.SpanWrappedHttpHandler.execute(SpanWrappedHttpHandler.java:87)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.grid.sessionqueue.NewSessionQueue.execute(NewSessionQueue.java:128)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.remote.http.Route.execute(Route.java:69)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.remote.http.Route$CombinedRoute.handle(Route.java:345)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.remote.http.Route.execute(Route.java:69)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.remote.http.Route$TemplatizedRoute.handle(Route.java:193)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.grid.sessionqueue.NewSessionQueue.lambda$new$0(NewSessionQueue.java:68)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.grid.sessionqueue.remote.RemoteNewSessionQueue.addToQueue(RemoteNewSessionQueue.java:104)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.remote.http.Filter.lambda$andFinally$1(Filter.java:55)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.grid.security.AddSecretFilter.lambda$apply$0(AddSecretFilter.java:40)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.remote.tracing.TracedHttpClient.execute(TracedHttpClient.java:54)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.remote.http.jdk.JdkHttpClient.execute(JdkHttpClient.java:366)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.remote.http.Filter.lambda$andFinally$1(Filter.java:55)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.remote.http.AddSeleniumUserAgent.lambda$apply$0(AddSeleniumUserAgent.java:42)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.remote.http.jdk.JdkHttpClient.execute0(JdkHttpClient.java:410)
2023-11-28T11:37:40+01:00	Driver info: driver.version: unknown
2023-11-28T11:37:40+01:00	System info: os.name: 'Linux', os.arch: 'amd64', os.version: '5.15.0-1036-azure', java.version: '11.0.20.1'
2023-11-28T11:37:40+01:00	Build info: version: '4.15.0', revision: '1d14b5521b'
2023-11-28T11:37:40+01:00	10:37:40.988 WARN [SeleniumSpanExporter$1.lambda$export$1] - org.openqa.selenium.TimeoutException: java.util.concurrent.TimeoutException
2023-11-28T11:37:40+01:00	Driver info: driver.version: unknown
2023-11-28T11:37:40+01:00	System info: os.name: 'Linux', os.arch: 'amd64', os.version: '5.15.0-1036-azure', java.version: '11.0.20.1'
2023-11-28T11:37:40+01:00	Build info: version: '4.15.0', revision: '1d14b5521b'
2023-11-28T11:37:40+01:00	10:37:40.987 WARN [SeleniumSpanExporter$1.lambda$export$1] - Unable to execute request: java.util.concurrent.TimeoutException
2023-11-28T11:37:40+01:00		... 43 more
2023-11-28T11:37:40+01:00		at org.openqa.selenium.remote.http.jdk.JdkHttpClient.execute0(JdkHttpClient.java:393)
2023-11-28T11:37:40+01:00		at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2021)
2023-11-28T11:37:40+01:00		at java.base/java.util.concurrent.CompletableFuture.timedGet(CompletableFuture.java:1886)
2023-11-28T11:37:40+01:00	Caused by: java.util.concurrent.TimeoutException
2023-11-28T11:37:40+01:00		at java.base/java.lang.Thread.run(Thread.java:829)
2023-11-28T11:37:40+01:00		at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
2023-11-28T11:37:40+01:00		at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
2023-11-28T11:37:40+01:00		at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
2023-11-28T11:37:40+01:00		at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.netty.server.SeleniumHandler.lambda$channelRead0$0(SeleniumHandler.java:44)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.remote.http.Filter$1.execute(Filter.java:63)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.remote.ErrorFilter.lambda$apply$0(ErrorFilter.java:44)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.remote.http.Filter$1.execute(Filter.java:63)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.remote.ErrorFilter.lambda$apply$0(ErrorFilter.java:44)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.remote.AddWebDriverSpecHeaders.lambda$apply$0(AddWebDriverSpecHeaders.java:35)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.remote.http.Route.execute(Route.java:69)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.remote.http.Route$CombinedRoute.handle(Route.java:345)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.remote.http.Route.execute(Route.java:69)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.remote.http.Route$CombinedRoute.handle(Route.java:345)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.remote.http.Route.execute(Route.java:69)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.remote.http.Route$CombinedRoute.handle(Route.java:345)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.remote.http.Route.execute(Route.java:69)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.remote.http.Route$CombinedRoute.handle(Route.java:345)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.remote.http.Route.execute(Route.java:69)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.remote.http.Route$NestedRoute.handle(Route.java:271)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.remote.http.Route.execute(Route.java:69)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.remote.http.Route$CombinedRoute.handle(Route.java:345)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.remote.http.Filter$1.execute(Filter.java:63)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.grid.web.EnsureSpecCompliantResponseHeaders.lambda$apply$0(EnsureSpecCompliantResponseHeaders.java:34)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.grid.router.Router.execute(Router.java:87)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.remote.http.Route.execute(Route.java:69)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.remote.http.Route$CombinedRoute.handle(Route.java:345)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.remote.http.Filter$1.execute(Filter.java:63)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.remote.tracing.SpanWrappedHttpHandler.execute(SpanWrappedHttpHandler.java:87)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.grid.sessionqueue.NewSessionQueue.execute(NewSessionQueue.java:128)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.remote.http.Route.execute(Route.java:69)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.remote.http.Route$CombinedRoute.handle(Route.java:345)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.remote.http.Route.execute(Route.java:69)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.remote.http.Route$TemplatizedRoute.handle(Route.java:193)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.grid.sessionqueue.NewSessionQueue.lambda$new$0(NewSessionQueue.java:68)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.grid.sessionqueue.remote.RemoteNewSessionQueue.addToQueue(RemoteNewSessionQueue.java:104)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.remote.http.Filter.lambda$andFinally$1(Filter.java:55)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.grid.security.AddSecretFilter.lambda$apply$0(AddSecretFilter.java:40)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.remote.tracing.TracedHttpClient.execute(TracedHttpClient.java:54)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.remote.http.jdk.JdkHttpClient.execute(JdkHttpClient.java:366)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.remote.http.Filter.lambda$andFinally$1(Filter.java:55)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.remote.http.AddSeleniumUserAgent.lambda$apply$0(AddSeleniumUserAgent.java:42)
2023-11-28T11:37:40+01:00		at org.openqa.selenium.remote.http.jdk.JdkHttpClient.execute0(JdkHttpClient.java:410)
2023-11-28T11:37:40+01:00	Driver info: driver.version: unknown
2023-11-28T11:37:40+01:00	System info: os.name: 'Linux', os.arch: 'amd64', os.version: '5.15.0-1036-azure', java.version: '11.0.20.1'
2023-11-28T11:37:40+01:00	Build info: version: '4.15.0', revision: '1d14b5521b'
2023-11-28T11:37:40+01:00	org.openqa.selenium.TimeoutException: java.util.concurrent.TimeoutException
2023-11-28T11:37:40+01:00	Driver info: driver.version: unknown
2023-11-28T11:37:40+01:00	System info: os.name: 'Linux', os.arch: 'amd64', os.version: '5.15.0-1036-azure', java.version: '11.0.20.1'
2023-11-28T11:37:40+01:00	Build info: version: '4.15.0', revision: '1d14b5521b'
2023-11-28T11:37:40+01:00	10:37:40.984 WARN [SpanWrappedHttpHandler.execute] - Unable to execute request: java.util.concurrent.TimeoutException
```

giuliohome commented:
And it looks like a session remains in the queue

![immagine](https://github.com/SeleniumHQ/selenium/assets/3272563/e6925a09-3cb7-4d0a-ab4f-427f91c99fa3)

![immagine](https://github.com/SeleniumHQ/selenium/assets/3272563/4c55f03a-fd55-4359-ae3c-50afff8ab9d0)

And it looks like a session remains in the queue. However, I **hope** this is due to issues **external** to the grid.Though, it remains strange for me that there was a **dangling** session after the corresponding tests were **completed** OK. At least, the error is a timeout, and it is **different** than the previous 'session not created.'

dummyuser992 commented:
if current issue seems like solved, can we release this fix?

dummyuser992 commented:
guys, who can make go/no go release this fix, since our team is experiencing this issue and we need to fix it?

diemol commented:
> guys, who can make go/no go release this fix, since our team is experiencing this issue and we need to fix it?

If you need this now, you can use Nightly: https://www.selenium.dev/downloads/#nightly

We will try to do a release next week if time and availability permits.

diemol commented:
@giuliohome those errors are not related to the issue, you see this error in the Node. I will mark those comments as off-topic.

giuliohome commented:
> @giuliohome those errors are not related to the issue, you see this error in the Node. I will mark those comments as off-topic.

I have forwarded your response to my colleague. From my perspective, I have communicated my stance on the matter - that this doesn't align with an open-source approach. I strongly advise against using such software in our company.

titusfortner commented:
what doesn't align with an open source approach? 

It appears that your recent comments do not apply to the solution provided by this PR. Please open a new issue if there is a problem that needs to be addressed.

giuliohome commented:
> what doesn't align with an open source approach?
> 
> It appears that your recent comments do not apply to the solution provided by this PR. Please open a new issue if there is a problem that needs to be addressed.

The way you assess that my comments do not apply and mark them as off-topic raises a concern. The point I'm making is precisely that my comments are closely tied to the ongoing discussion.

Specifically, regarding the use of Process vs. ExternalProcess, and the feedback requested just before them. 


More broadly, the issue extends to the lack of clarification in your statement:

> those errors are not related to the issue, you see this error in the Node.

This statement is rather brief and could be interpreted as somewhat dismissive. To provide a clearer context, I'd like to emphasize that I observed the error in the router log, seemingly linked to an exception when saving the process output - a matter discussed previously. Even if there's a misunderstanding on my part, the communication doesn't seem conducive to a collaborative approach that values understanding users' perspectives (from a devrel communication perspective, it's crucial to ensure a more open and receptive tone, even when closing a comment).

diemol commented:
> > what doesn't align with an open source approach?
> > It appears that your recent comments do not apply to the solution provided by this PR. Please open a new issue if there is a problem that needs to be addressed.
> 
> The way you assess that my comments do not apply and mark them as off-topic raises a concern. The point I'm making is precisely that my comments are closely tied to the ongoing discussion.
> 
> Specifically, regarding the use of Process vs. ExternalProcess, and the feedback requested just before them.
> 
> More broadly, the issue extends to the lack of clarification in your statement:
> 
> > those errors are not related to the issue, you see this error in the Node.
> 
> This statement is rather brief and could be interpreted as somewhat dismissive. To provide a clearer context, I'd like to emphasize that I observed the error in the router log, seemingly linked to an exception when saving the process output - a matter discussed previously. Even if there's a misunderstanding on my part, the communication doesn't seem conducive to a collaborative approach that values understanding users' perspectives (from a devrel communication perspective, it's crucial to ensure a more open and receptive tone, even when closing a comment).

I replied that because Selenium Manager is being used only in the Node, and the issue is about that, not the Router, where no external processes are executed, everything is already in the JAR file. If you are seeing something similar that requires a separate issue with enough context, let us know how to troubleshoot it. That is why I marked them as off-topic.

giuliohome commented:
The situation is much clearer now. It turns out I was initially selecting the wrong pod in Grafana. For instance, the error I observed from the Node pod is as follows: 

```
2023-11-29 12:34:50.418 | at java.base/java.lang.Thread.run(Thread.java:829) |  
-- | -- | --
  |   | 2023-11-29 12:34:50.418 | at org.openqa.selenium.os.ExternalProcess$Builder.lambda$start$0(ExternalProcess.java:209) |  
  |   | 2023-11-29 12:34:50.418 | at java.base/java.io.InputStream.transferTo(InputStream.java:704) |  
  |   | 2023-11-29 12:34:50.418 | at java.base/java.io.BufferedInputStream.read(BufferedInputStream.java:342) |  
  |   | 2023-11-29 12:34:50.418 | at java.base/java.io.BufferedInputStream.getBufIfOpen(BufferedInputStream.java:176) |  
  |   | 2023-11-29 12:34:50.418 | java.io.IOException: Stream closed |  
  |   | 2023-11-29 12:34:50.418 | 11:34:50.416 WARN [ExternalProcess$Builder.lambda$start$0] - failed to copy the output of process 15191
```

This error seems somehow related to the ongoing discussion.

Additionally, during the same timeframe, the Kubernetes Ingress Controller reported lines referring to both the Router and the download part of our test:

```
2023-11-29 12:34:50.108 | {   "namespace": "test-automation",   "ingress_name": "selenium-ingress",   "service_name": "selenium-router",   "service_port": "4444",   "request_id": "a1321224cf10b6004290e9bd2c2a7645",   "request_length": "471",   "remote_addr": "10.65.131.136",   "remote_port": "45687",   "time_iso8601": "2023-11-29T11:34:50+00:00",   "request_uri": "/wd/hub/session/7a6183da89044bb24d29a58fe1df7fc5/se/files"
-- | --
```

dummyuser992 commented:
can we merge it and release?

diemol commented:
4.16.1 is out, are you still seeing the issue with this version?

titusfortner commented:
> This statement is rather brief and could be interpreted as somewhat dismissive

Please be careful about reading into things in responses here. The Selenium team is composed of people from many countries and cultures and languages. You will see a variety of tones and terseness in the conversations in our Slack and in our GitHub issues.

We get a lot of new issues every day and most of the work is the thankless job of trying to differentiate between things that need to be fixed and things that don't need to be fixed, and things that help us determine between those conditions and things that do not. If something can get put in a "not helpful" bucket, we're going to quickly put it there so we can sift through the next pile of information.

giuliohome commented:
> We get a lot of new issues every day and most of the work is the thankless job of trying to differentiate between things that need to be fixed and things that don't need to be fixed

That's also my job, though in a different context. Meanwhile I understood what was the cause of the message that has been put off-topic here and indeed it has been thanks to a colleague of mine that pointed me to the problem by asking me to solve an issue seen by a dev team working for us. So again, my conclusion is that before differentiating things must be well understood and often the communication among different teams is key. 

Back to the main topic, here now I don't understand why someone wants to merge this if the underlying issue seems to be already fixed in a quite different way:
> now the browser driver path is set via properties and with that, Selenium Manager is avoided

titusfortner commented:
>  don't understand why someone wants to merge this if the underlying issue seems to be already fixed in a quite different way.

The issue is that while Selenium Manager should work on Docker, it shouldn't get used on Docker. So we added the code to keep it from getting used. But the same thing that caused the problem we saw on Docker may exist in non-docker environments, and we may need this PR to address it, so we're leaving it open.

bonigarcia commented:
I believe this PR is no longer needed. I'm closing it.
